### PR TITLE
SLEAP read user labels

### PR DESCRIPTION
**Pull request by lochhh**
## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**
This PR enables `load_poses._load_from_sleap_labels_file` to load pose tracks from a SLEAP .slp file consisting of either:
- user-labelled instances only,
- predicted instances only, or
- a mix of user-labelled and predicted instances

**What does this PR do?**
This PR adds:
- `load_poses._sleap_labels_to_numpy`, a function that converts SLEAP `Labels` to a NumPy array, adapted from [`Labels.numpy()`](https://github.com/talmolab/sleap-io/blob/9a04f7feac0397d7d7bed9fb79f205678cccf9f8/sleap_io/model/labels.py#L92-L178) from the [`sleap_io`](https://github.com/talmolab/sleap-io) package, but prioritises user-labelled instances over predicted instances when the .slp file consists of a mix of user-labelled and predicted instances. This behaviour mirrors [SLEAP's approach when exporting ".h5" analysis files](https://github.com/talmolab/sleap/blob/v1.3.3/sleap/info/write_tracking_h5.py#L129-L150), so that the loading a .slp file in _movement_ returns the same dataset as the corresponding SLEAP analysis .h5 file (i.e., exported from the same .slp file). 
- new test cases for 
  - checking that loading pairs of .h5 and .slp files returns the same dataset, regardless of whether the .slp file consists of user-labelled instances only, predicted instances only, or a mix of both
  - converting SLEAP datasets to DeepLabCut-style pandas DataFrames
  - converting SLEAP datasets to DeepLabCut file formats 

## References
This PR addresses #84 

## How has this PR been tested?
Tests have been written and run for the new functions added.

## Does this PR require an update to the documentation?
Docstrings have been updated accordingly.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: movement/io/load_poses.py [hunk 1]
```diff
@@ -109,7 +111,7 @@ def from_sleap_file(
     preferred format for loading pose tracks from SLEAP into *movement*.
 
     You can also try directly loading the ".slp" file, but this feature is
-    experimental and doesnot work in all cases. If the ".slp" file contains
+    experimental and does not work in all cases. If the ".slp" file contains
     both user-labeled and predicted instances, only the predicted ones will be
     loaded. If there are multiple videos in the file, only the first one will
     be used.
```
Comment (niksirbi) [inline movement/io/load_poses.py:117]: I think this section of the docstring could be updated. This is the user's enty point for SLEAP files, so we want to make sure this docstring is solid.

```suggestion
    You can also try directly loading the ".slp" file, but this feature
    does not work in all cases. If the ".slp" file contains only predicted
    intances, they will be imported alongside their point-wise confidence scores.
    Alternatively, if it contains only user-labeled instances, they will be
    loaded and assigned fixed point-wise confidence scores of 1.0.
    Lastly, if a mix of both user-labeled and predicted instances are present,
    only the predicted ones will be loaded.
    If there are multiple videos in the file, only the first one will be used.
```

Feel free to edit if my interpretation of the expected behaviour is incorrect

File: movement/io/load_poses.py [hunk 2]
```diff
@@ -284,16 +301,101 @@ def _load_from_sleap_labels_file(
...
+        video of the SLEAP `Labels` object. It is primarily meant
+        to be used with `Labels` containing only user-labeled instances.
+        If `Labels` contains predicted instances only, this function
+        will overwrite all point scores to 1.0. If `Labels` contains
+        both user-labeled and predicted instances, the returned NumPy
+        array will contain the last occurrence of each tracked instance
+        in each frame.
+
+        This function is adapted from `Labels.numpy()` from the
+        `sleap_io` package [1]_ to allow user-labeled instances.
```
Comment (niksirbi) [inline movement/io/load_poses.py:353]: There is an extra level of indentation for some reason
```suggestion
    This function only considers SLEAP instances in the first
    video of the SLEAP `Labels` object. It is primarily meant
    to be used with `Labels` containing only user-labeled instances.
    If `Labels` contains predicted instances only, this function
    will overwrite all point scores to 1.0. If `Labels` contains
    both user-labeled and predicted instances, the returned NumPy
    array will contain the last occurrence of each tracked instance
    in each frame.

    This function is adapted from `Labels.numpy()` from the
    `sleap_io` package [1]_ to allow user-labeled instances.
```

**Review by niksirbi — APPROVED** [review]
Great job @lochhh!

I have two general comments on this, apart from the specifics:

- I think we need to explicitly test for the case that a .slp file contains both predicted and user-labeled instances and make sure that we load the expected things in that case (according to what you describe in the docstring of `_sleap_user_labels_to_numpy()`)
- The `load_poses.py` is starting to get a bit long. How about refactoring the private `_` functions into a separate `movement.io.utils.py` file? If you agree with this, let's open an issue to track this (we should do it in a different PR, after this one is merged).

File: movement/io/load_poses.py [hunk 3]
```diff
@@ -227,6 +229,12 @@ def _load_from_sleap_analysis_file(
     -------
     movement.io.tracks_validators.ValidPoseTracks
         The validated pose tracks and confidence scores.
+
+    Notes
+    -----
+    If the point-level prediction scores (i.e. confidence) in the SLEAP
+    analysis file are all NaNs, this function assumes that the pose
+    tracks are user-labeled, and will assign a fixed score of 1.0.
```
Comment (niksirbi) [inline movement/io/load_poses.py:237]: I agree with making this assumption, but could be risky to do this "silently". Should we log a warning when we do this? Both here and for the labels file.

Comment (lochhh) [reply movement/io/load_poses.py:237]: Good idea! Will do that.

File: tests/test_unit/test_load_poses.py [hunk 4]
```diff
@@ -2,13 +2,32 @@
...
+        )
+        return read_labels(slp_file)
+
+    @pytest.fixture(scope="module")
+    def sleap_predicted_labels(self):
+        """Return a SLEAP `Labels` object containing only
+        predicted instances.
+        """
+        slp_file = POSE_DATA.get("SLEAP_single-mouse_EPM.predictions.slp")
+        return read_labels(slp_file)
```
[the remark this answers is not in the document]
Comment (lochhh) [reply tests/test_unit/test_load_poses.py]: I think this concerns a separate issue. For simplicity, let's consider a single-animal frame here.
Possible scenarios leading to a mix of Predicted and User-labelled Instances in a `sleap.instance.LabeledFrame`:
- case 1 - user "fixes" incorrectly predicted individual A: 
  `[PredictedInstance(A)]` &rarr; `[PredictedInstance(A), Instance(A)]` 
- case 2 - user "adds" missing individual A: 
  `[]` &rarr; `[Instance(A)]`  

In either case, the current behaviour using `sleap-io`'s [`Labels.numpy()`](https://github.com/talmolab/sleap-io/blob/9a04f7feac0397d7d7bed9fb79f205678cccf9f8/sleap_io/model/labels.py#L92) is to ignore user-labelled instances:
- loading `case1.slp` in _movement_ returns `[PredictedInstance(A)]`
- loading `case2.slp` in _movement_ returns `[]`

However, if the user exports `case1.slp` as `case1.h5`, and `case2.slp` as `case2.h5`, SLEAP's [`write_tracking_h5` prioritises the user-labelled instances](https://github.com/talmolab/sleap/blob/9eeba979b7ce32400809723b668a8bd41e09d973/sleap/info/write_tracking_h5.py#L138-L141), i.e., if user-labelled instance is present, select user-labelled instance, otherwise select predicted instance. Hence:
- loading `case1.h5` in _movement_ returns `[Instance(A)]` with NaN confidence
- loading `case2.h5` in _movement_ returns `[Instance(A)]` with NaN confidence

I guess we should handle .slp files with mixed instance types the same way SLEAP does in `write_tracking_h5`, so that loading `.slp` and `.h5`  in _movement_ return the same dataset. And perhaps, for consistency, we should leave confidence scores for user-labels-only files as NaNs, rather than setting these as 1.0... 

Wrt test data, I think it's more straightforward to add a pair of .h5 and .slp files to our sample datasets.

Wdyt?

[the remark this answers is not in the document]
Comment (niksirbi) [reply tests/test_unit/test_load_poses.py]: I thought about it, and I think what you propose here makes sense and aligns better with the expectations of users. If I correct a prediction (thereby, turning it into a user-labelled instance), I expect it to be prioritised for going into the output, like `write_tracking_h5` does.

Moreover, as you point out, it makes sense for the imported data to be the same, whether it came from `.slp` or `.h5`. This means that we should indeed duplicate the functionality of `write_tracking_h5`. Since we cannot depend on `sleap`, and `sleap-io` doesn't have it, we should write our own implementation for it.

There are two issues I foresee with this approach:
1. We have to keep an eye on the `write_tracking_h5` in `sleap`. If it's behaviour changes significantly, we may need to update our own version. But I hope they won't do that, because if they do, we'll have to ship different data loaders for different versions of `sleap`.  But I guess this is the case already, we have to stay up-to-date with developments in `deeplabcut` and `sleap`.
2. As you mention, we probably have to mark the user-labelled instances with NaN confidence. This will not be a problem if we assume that user-labelling is the only process that can generate NaNs in `.slp` and `.h5` files. If this assumption holds, we can treat NaNs coming from `sleap` as having the highest confidence during preprocessing. If this assumption is not true, we need another way of keeping track of which instances are user-labelled (maybe an additional boolean `DataArray` or a list of the indices of  user-labelled instances stored in the `Dataset`'s metadata `attrs`.

In summary, I think you should go ahead with implementing this plan in this PR, and we should come to a decision on how to handle NaNs. We definitely need to keep information on which predictions we can "trust",  because it will come in handy during data cleaning.

**Review by niksirbi — APPROVED** [review]
That's very good @lochhh !
I thinks this is ready to go, apart from one tiny comment I left.

We should merge this ahead of [PR 83](https://github.com/neuroinformatics-unit/movement/pull/83), because some of the tests I worte for that one will conflict with the ones here. After we merge this one, I'll merge/rebase PR 83, fix conflicts and merge that one too.

File: movement/io/load_poses.py [hunk 5]
```diff
@@ -108,11 +109,8 @@ def from_sleap_file(
     "Export Analysis HDF5…" from the "File" menu) [1]_. This is the
     preferred format for loading pose tracks from SLEAP into *movement*.
 
-    You can also try directly loading the ".slp" file, but this feature is
-    experimental and doesnot work in all cases. If the ".slp" file contains
-    both user-labeled and predicted instances, only the predicted ones will be
-    loaded. If there are multiple videos in the file, only the first one will
-    be used.
+    You can also directly load the ".slp" file. However, if the file contains
+    multiple videos, only the pose tracks from the first video will be loaded.
```
[the remark this answers is not in the document]
Comment (lochhh) [reply movement/io/load_poses.py:113]: Good point! Will add that.
